### DISCUSSION TITLE
revalidateIfStale false and keepPreviousData true will break pages with fallback data

### DISCUSSION THREAD

**Discussion by ddhp** [vercel/swr#2489] [Q&A]
# Bug report

## Description / Observed Behavior
If setting `revalidateIfStale` false and `keepPreviousData` true, changing between pages with fallback data wouldn't work.

## Expected Behavior
returning fallback data from SSR

## Repro Steps / Code Example

[Reproduce codesandbox](https://codesandbox.io/p/sandbox/musing-hawking-8ie7jw?file=%2Fpages%2F%5Bpid%5D.tsx&selection=%5B%7B%22endColumn%22%3A9%2C%22endLineNumber%22%3A32%2C%22startColumn%22%3A9%2C%22startLineNumber%22%3A32%7D%5D): navigating between `/a` and `/b` and swr hook always return old data.

I think it's from here
https://github.com/vercel/swr/blob/f7004ac37b2b06111043bb6f262f80fb4d5cc381/core/use-swr.ts#L206-L216
when `revalidateIfStale` is false, cachedData wouldn't be updated from fallback and when `keepPreviousData` is true, laggyRef data will be returned.

if modifies returnedData to
```js
const returnedData = data ? data : keepPreviousData
    ? isUndefined(cachedData)
      ? laggyDataRef.current
      : cachedData
```
will solve the problem but I don't know if this is the expect behavior.

Please help to clarify, thanks!

## Additional Context
tested on
swr v2.0.3
next v13.1.6

**Comment by koba04** [comment]
@ddhp 
Thank you! I've confirmed the behavior. This is because we prioritize `keepPreviousData` over `fallback`. Why do you enable `keepPreviousData` even if you have `fallback`?

**Comment by ddhp** [comment]
Hi @koba04, thanks for your reply!

I think we want to keep the original data when the key changes and meanwhile the page/component supports SSR.
Is it like what this diagram explains?
https://swr.vercel.app/docs/advanced/understanding#key-change--previous-data--fallback

**Comment by koba04** [comment]
The dialog is a case that has the fallback data only for the initial fetch, so it's different from this.

**Comment by ddhp** [comment]
Hmm.. I can't get it here.

For our scenario, we have a fallback data for the initial fetch, then there is a key change caused by changing page but we can't see the hook getting revalidated. 

From my understanding, `keepPreviousData` will keep the previous data of the previous key until the data for the new key is revalidated; And it should revalidate even when we set `revalidateIfStale` to false since the key has changed.
